File: mobile-app/src/lib/parsers/cdapParser.ts
/**
 * CDAP Parser — Full JavaScript port of backend cdap_parser.py
 *
 * Extracts units with Part 1/Part 2 topics from PDF, DOCX, and Excel files.
 *
 * CDAP (Course Delivery and Assessment Plan) Structure:
 *   - Units/Modules
 *   - Each unit has Part 1 and Part 2 topics
 *   - Topics may have subtopics
 *
 * All processing is local — no network calls, no backend dependency.
 */

import * as XLSX from 'xlsx';
import { extractTextFromPDF } from '../pdfExtractor';
import type { CDAPUnit, CDAPTopicEntry } from '../../types';

// Dynamically import mammoth only when needed
let mammothModule: any = null;
async function getMammoth() {
  if (!mammothModule) {
    mammothModule = await import('mammoth');
  }
  return mammothModule;
}

// Re-export types for consumers
export type { CDAPUnit, CDAPTopicEntry };

// Helper to extract topic string from union type
function getTopicStr(t: string | CDAPTopicEntry): string {
  return typeof t === 'string' ? t : t.topic;
}

export class CDAPParser {
  // ======================== Public Entry Point ========================

  /**
   * Parse CDAP from any supported file format.
   */
  static async parseFile(file: File): Promise<CDAPUnit[]> {
    const ext = file.name.split('.').pop()?.toLowerCase() || '';

    if (ext === 'pdf') {
      return this.parsePDF(file);
    } else if (['docx', 'doc'].includes(ext)) {
      return this.parseDOCX(file);
    } else if (['xlsx', 'xls', 'csv'].includes(ext)) {
      return this.parseExcel(file);
    } else {
      throw new Error(`Unsupported file type: ${ext}`);
    }
  }

  // ======================== Format-Specific Parsers ========================

  /**
   * Parse PDF CDAP: extract text + tables, parse structure.
   */
  static async parsePDF(file: File): Promise<CDAPUnit[]> {
    try {
      console.log('[CDAPParser] Extracting PDF text via pdfjs-dist...');
      const text = await extractTextFromPDF(file);
      console.log(`[CDAPParser] Extracted ${text.length} chars from PDF`);

      // For PDFs, first try text-based CDAP structure extraction
      const result = this.extractCDAPStructure(text);
      if (result.length > 0) return result;

      // Fallback: try table-row parsing on split lines
      const lines = text.split('\n').map(l => l.split(/\s{2,}|\t/));
      const tableResult = this.parseTableRows(lines);
      return tableResult;
    } catch (error) {
      console.error('[CDAPParser] PDF parse error:', error);
      return [];
    }
  }

  /**
   * Parse DOCX CDAP: extract text with mammoth, parse structure.
   */
  static async parseDOCX(file: File): Promise<CDAPUnit[]> {
    try {
      console.log('[CDAPParser] Extracting DOCX text via mammoth...');
      const mammoth = await getMammoth();
      const arrayBuffer = await file.arrayBuffer();
      const result = await mammoth.extractRawText({ arrayBuffer });
      const text = result.value || '';
      console.log(`[CDAPParser] Extracted ${text.length} chars from DOCX`);
      return this.extractCDAPStructure(text);
    } catch (error) {
      console.error('[CDAPParser] DOCX parse error:', error);
      return [];
    }
  }

  /**
   * Parse Excel CDAP: structured column-based parsing.
   *
   * Expected structure (e.g., OS CDAP.xlsx):
   *   Column A: UNIT number (1.0, 2.0, etc.)
   *   Column B: SYLLABUS (UNIT NAME)
   *   Column C: OUTCOME (CO)
   *   Column D: PART NO. (1.0 or 2.0)
   *   Column E: TOPICS TO BE COVERED
   *   Column F: SUB TOPICS
   *   Column G: BT LEVEL
   */
  static async parseExcel(file: File): Promise<CDAPUnit[]> {
    try {
      const buffer = await file.arrayBuffer();
      const workbook = XLSX.read(buffer, { type: 'array' });
      const sheetName = workbook.SheetNames[0];
      const worksheet = workbook.Sheets[sheetName];
      const data = XLSX.utils.sheet_to_json(worksheet, { header: 1 }) as any[][];

      return this.parseTableRows(data);
    } catch (error) {
      console.error('[CDAPParser] Excel parse error:', error);
      return [];
    }
  }

  // ======================== Table Row Parser (ported from Python) ========================

  /**
   * Parse structured table rows.
   * Detects columns dynamically from header.
   *
   * Ported from cdap_parser.py → _parse_excel() / _parse_table_rows()
   */
  private static parseTableRows(rows: any[][]): CDAPUnit[] {
    if (!rows || rows.length === 0) return [];

    const units: Record<number, CDAPUnit> = {};
    let currentUnit: number | null = null;
    let currentUnitName: string | null = null;

    // Find header row (within first 5 rows)
    let headerRowIdx = 0;
    for (let idx = 0; idx < Math.min(5, rows.length); idx++) {
      const rowValues = (rows[idx] || []).map(c => (c != null ? String(c).trim().toUpperCase() : ''));
      if (rowValues.some(v => v.includes('UNIT') || v.includes('PART') || v.includes('TOPIC'))) {
        headerRowIdx = idx;
        break;
      }
    }

    // Detect column indices from header
    const colIndices = { unit: 0, unit_name: 1, part: 3, topic: 4, subtopic: 5 };
    if (headerRowIdx < rows.length) {
      const header = rows[headerRowIdx] || [];
      header.forEach((cell: any, idx: number) => {
        if (!cell) return;
        const cellUpper = String(cell).toUpperCase().replace(/\n/g, ' ');

        if (cellUpper.includes('UNIT') && !cellUpper.includes('NAME') && idx < 2) {
          colIndices.unit = idx;
        } else if (cellUpper.includes('SYLLABUS') || (cellUpper.includes('UNIT') && cellUpper.includes('NAME'))) {
          colIndices.unit_name = idx;
        } else if (cellUpper.includes('PART')) {
          colIndices.part = idx;
        } else if (cellUpper.includes('TOPIC') && !cellUpper.includes('SUB')) {
          colIndices.topic = idx;
        } else if (cellUpper.includes('SUB') && cellUpper.includes('TOPIC')) {
          colIndices.subtopic = idx;
        }
      });
    }

    // Parse data rows
    for (let i = headerRowIdx + 1; i < rows.length; i++) {
      const row = rows[i];
      if (!row || row.every(cell => cell == null)) continue;

      const rowValues = row.map(c => (c != null ? String(c).trim().replace(/\n/g, ' ') : ''));

      // Skip invalid rows
      if (rowValues[0]?.includes('#REF!')) continue;

      let partNum: number | null = null;

      // Check Unit column
      const unitCol = colIndices.unit;
      if (rowValues.length > unitCol && rowValues[unitCol]) {
        const unitVal = rowValues[unitCol];
        const skipHeaders = ['UNIT', 'UNIT NO', 'UNIT NO.', 'UNIT NUMBER', 'NONE', ''];
        if (!skipHeaders.includes(unitVal.toUpperCase())) {
          const uMatch = unitVal.match(/(?:unit|module)?\s*[:\-]?\s*([\d.]+)/i);
          if (uMatch) {
            currentUnit = Math.floor(parseFloat(uMatch[1]));
            const unitNameCol = colIndices.unit_name;
            if (rowValues.length > unitNameCol && rowValues[unitNameCol]) {
              currentUnitName = rowValues[unitNameCol];
            }
          }
        }
      }

      if (currentUnit === null) currentUnit = 1;

      // Get Part Number
      const partCol = colIndices.part;
      if (rowValues.length > partCol && rowValues[partCol]) {
        const partVal = rowValues[partCol];
        const skipPartHeaders = ['PART NO', 'PART NO.', 'PART NUMBER', 'PART', 'NONE', ''];
        if (!skipPartHeaders.includes(partVal.toUpperCase())) {
          if (['1', '1.0', 'Part 1', 'Part-1', 'I', 'PART I'].includes(partVal)) {
            partNum = 1;
          } else if (['2', '2.0', 'Part 2', 'Part-2', 'II', 'PART II'].includes(partVal)) {
            partNum = 2;
          } else {
            const pMatch = partVal.match(/[\d.]+/);
            if (pMatch) partNum = Math.floor(parseFloat(pMatch[0]));
          }
        }
      }

      // Get Topic
      const topicCol = colIndices.topic;
      const topicsToAdd: string[] = [];
      if (rowValues.length > topicCol && rowValues[topicCol]) {
        const topicVal = rowValues[topicCol];
        if (!topicVal.toUpperCase().includes('TOPIC') || topicVal.length > 30) {
          if (topicVal.length > 3 && topicVal.toUpperCase() !== 'NONE') {
            topicsToAdd.push(topicVal);
          }
        }
      }

      // Initialize unit if needed
      if (!(currentUnit in units)) {
        const unitName = currentUnitName || `Unit ${currentUnit}`;
        units[currentUnit] = {
          unit_number: currentUnit,
          unit_name: unitName,
          part1_topics: [],
          part2_topics: [],
        };
      } else if (currentUnitName && units[currentUnit].unit_name === `Unit ${currentUnit}`) {
        units[currentUnit].unit_name = currentUnitName;
      }

      // Add topics to the correct part
      for (const topicText of topicsToAdd) {
        if (!topicText) continue;

        // Get subtopic if available
        const subtopics: string[] = [];
        const subtopicCol = colIndices.subtopic;
        if (rowValues.length > subtopicCol && rowValues[subtopicCol]) {
          const subtopicVal = rowValues[subtopicCol].trim();
          if (subtopicVal && subtopicVal.length > 3 && !subtopicVal.toUpperCase().startsWith('SUB')) {
            subtopics.push(subtopicVal);
          }
        }

        const targetPart = partNum || 1;
        const targetList = targetPart === 1 ? units[currentUnit].part1_topics : units[currentUnit].part2_topics;

        // Check for duplicates
        const existing = targetList.find(t => getTopicStr(t) === topicText);
        if (existing && typeof existing !== 'string') {
          for (const st of subtopics) {
            if (!existing.subtopics.includes(st)) {
              existing.subtopics.push(st);
            }
          }
        } else {
          targetList.push({ topic: topicText, subtopics });
        }
      }
    }

    return Object.values(units);
  }

  // ======================== Text-Based CDAP Parser ========================

  /**
   * Extract CDAP structure from raw text.
   * Identifies units and Part 1/Part 2 sections.
   *
   * Ported from cdap_parser.py → _extract_cdap_structure()
   */
  private static extractCDAPStructure(text: string): CDAPUnit[] {
    const units: CDAPUnit[] = [];
    let currentUnit: CDAPUnit | null = null;
    let currentPart: number | null = null;

    const lines = text.split('\n');

    for (const rawLine of lines) {
      const line = rawLine.trim();
      if (!line) continue;

      // Detect unit/module headers
      let unitMatch = line.match(/(?:unit|module)\s*[-:]?\s*([IVX\d]+)(?:\s*[-:]\s*(.+))?/i);

      // Alternative: "1 Introduction..." or "1 | Introduction..."
      if (!unitMatch) {
        const simpleMatch = line.match(/^(\d+)(?:\s+|(?:\s*[|]\s*))([A-Za-z].+)/);
        if (simpleMatch) {
          const uNum = parseInt(simpleMatch[1], 10);
          if (uNum >= 1 && uNum <= 10) {
            unitMatch = simpleMatch;
          }
        }
      }

      if (unitMatch) {
        const unitNum = this.romanToInt(unitMatch[1]);
        const unitName = unitMatch[2]?.trim() || `Unit ${unitNum}`;

        currentUnit = {
          unit_number: unitNum,
          unit_name: unitName,
          part1_topics: [],
          part2_topics: [],
        };
        units.push(currentUnit);
        currentPart = null;
        continue;
      }

      // Detect Part 1/Part 2 headers
      if (/part\s*[-:]?\s*1|part\s*i(?:\s|$)/i.test(line)) {
        currentPart = 1;
        continue;
      }
      if (/part\s*[-:]?\s*2|part\s*ii(?:\s|$)/i.test(line)) {
        currentPart = 2;
        continue;
      }

      // Add topics to current part
      if (currentUnit && currentPart && this.isValidCDAPTopic(line)) {
        const topic = this.cleanCDAPTopic(line);
        if (topic) {
          const topicEntry: CDAPTopicEntry = { topic, subtopics: [] };
          if (currentPart === 1) {
            if (!currentUnit.part1_topics.some(t => getTopicStr(t) === topic)) {
              currentUnit.part1_topics.push(topicEntry);
            }
          } else {
            if (!currentUnit.part2_topics.some(t => getTopicStr(t) === topic)) {
              currentUnit.part2_topics.push(topicEntry);
            }
          }
        }
      }
    }

    return units;
  }

  // ======================== Validation & Cleaning ========================

  /**
   * Check if text is a valid CDAP topic.
   */
  private static isValidCDAPTopic(text: string): boolean {
    text = text.trim();
    if (text.length < 5) return false;
    if (/^[\d.]+$/.test(text)) return false;
    if (['part 1', 'part 2', 'part i', 'part ii', 'topics'].includes(text.toLowerCase())) return false;
    return true;
  }

  /**
   * Clean a CDAP topic string.
   */
  private static cleanCDAPTopic(text: string): string {
    // Remove leading numbers/bullets
    text = text.replace(/^[\d.•\-–]+\s*/, '');
    // Collapse whitespace
    text = text.replace(/\s+/g, ' ');
    return text.trim();
  }

  // ======================== Utilities ========================

  /**
   * Convert Roman numeral or digit string to integer.
   */
  private static romanToInt(s: string): number {
    s = s.trim().toUpperCase();
    if (/^\d+$/.test(s)) return parseInt(s, 10);

    const roman: Record<string, number> = { I: 1, V: 5, X: 10 };
    let result = 0;
    let prev = 0;
    for (const char of [...s].reverse()) {
      const val = roman[char] || 0;
      if (val < prev) {
        result -= val;
      } else {
        result += val;
      }
      prev = val;
    }
    return result > 0 ? result : 1;
  }
}

File: mobile-app/android/src/main/python/cdap_parser.py
import re
from typing import List, Dict, Optional
import pdfplumber
from docx import Document
import openpyxl
import os


class CDAPParser:
    """Parse CDAP (Course Delivery and Assessment Plan) from PDF, DOCX, or Excel files.
    
    CDAP Structure:
    - Units/Modules
    - Each unit has Part 1 and Part 2 topics
    """
    
    def parse_file(self, file_path: str) -> List[Dict]:
        """Parse CDAP file and extract units with Part 1/Part 2 topics"""
        ext = os.path.splitext(file_path)[1].lower()
        
        if ext == '.pdf':
            return self._parse_pdf(file_path)
        elif ext in ['.docx', '.doc']:
            return self._parse_docx(file_path)
        elif ext in ['.xlsx', '.xls']:
            return self._parse_excel(file_path)
        else:
            raise ValueError(f"Unsupported file type: {ext}")
    
    def _parse_pdf(self, file_path: str) -> List[Dict]:
        """Parse PDF CDAP - handles both table-based and text-based PDFs"""
        try:
            all_table_rows = []
            full_text = ""
            
            with pdfplumber.open(file_path) as pdf:
                for page in pdf.pages:
                    # Extract tables first (preferred for structured CDAPs)
                    tables = page.extract_tables()
                    if tables:
                        for table in tables:
                            for row in table:
                                if row and any(cell for cell in row):
                                    all_table_rows.append(row)
                    
                    # Also extract text for fallback
                    text = page.extract_text()
                    if text:
                        full_text += text + "\n"
            
            # If we have table data, use table-based parsing (similar to Excel)
            if all_table_rows:
                result = self._parse_table_rows(all_table_rows)
                if result:
                    return result
            
            # Fallback to text-based parsing
            return self._extract_cdap_structure(full_text)
        except Exception as e:
            print(f"PDF parse error: {e}")
            import traceback
            traceback.print_exc()
            return []
    
    def _parse_table_rows(self, rows: List) -> List[Dict]:
        """Parse table rows (from PDF or other sources) with same logic as Excel"""
        if not rows:
            return []
        
        units = {}
        current_unit = None
        current_unit_name = None
        
        # Find header row
        header_row_idx = 0
        for idx, row in enumerate(rows[:5]):
            row_values = [str(c).strip().upper() if c else '' for c in row]
            if any('UNIT' in v or 'PART' in v or 'TOPIC' in v for v in row_values):
                header_row_idx = idx
                break
        
        # Detect column indices from header
        col_indices = {'unit': 0, 'unit_name': 1, 'part': 3, 'topic': 4, 'subtopic': 5}
        if header_row_idx < len(rows):
            header = rows[header_row_idx]
            for idx, cell in enumerate(header):
                if cell:
                    cell_upper = str(cell).upper().replace('\n', ' ')
                    if 'UNIT' in cell_upper and 'NAME' not in cell_upper and idx < 2:
                        col_indices['unit'] = idx
                    elif 'SYLLABUS' in cell_upper or ('UNIT' in cell_upper and 'NAME' in cell_upper):
                        col_indices['unit_name'] = idx
                    elif 'PART' in cell_upper:
                        col_indices['part'] = idx
                    elif 'TOPIC' in cell_upper and 'SUB' not in cell_upper:
                        col_indices['topic'] = idx
                    elif 'SUB' in cell_upper and 'TOPIC' in cell_upper:
                        col_indices['subtopic'] = idx
        
        # Parse data rows
        for row in rows[header_row_idx + 1:]:
            if not row or all(cell is None for cell in row):
                continue
            
            row_values = [str(c).strip().replace('\n', ' ') if c is not None else '' for c in row]
            
            # Skip rows with #REF! or invalid data
            if row_values and '#REF!' in row_values[0]:
                continue
            
            # Initialize variables for this row
            part_num = None
            
            # Check column for Unit Number
            unit_col = col_indices['unit']
            if len(row_values) > unit_col and row_values[unit_col]:
                unit_val = row_values[unit_col]
                if unit_val.upper() not in ['UNIT', 'UNIT NO', 'UNIT NO.', 'UNIT NUMBER', 'NONE', '']:
                    u_match = re.search(r'(?:unit|module)?\s*[:\-]?\s*([\d.]+)', unit_val, re.IGNORECASE)
                    if u_match:
                        unit_num = int(float(u_match.group(1)))
                        current_unit = unit_num
                        unit_name_col = col_indices['unit_name']
                        if len(row_values) > unit_name_col and row_values[unit_name_col]:
                            current_unit_name = row_values[unit_name_col]
            
            if current_unit is None:
                current_unit = 1
            
            # Get Part Number
            part_col = col_indices['part']
            if len(row_values) > part_col and row_values[part_col]:
                part_val = row_values[part_col]
                if part_val.upper() not in ['PART NO', 'PART NO.', 'PART NUMBER', 'PART', 'NONE', '']:
                    if part_val in ['1', '1.0', 'Part 1', 'Part-1', 'I', 'PART I']:
                        part_num = 1
                    elif part_val in ['2', '2.0', 'Part 2', 'Part-2', 'II', 'PART II']:
                        part_num = 2
                    else:
                        p_match = re.search(r'[\d.]+', part_val)
                        if p_match:
                            part_num = int(float(p_match.group()))
            
            # Get Topic
            topic_col = col_indices['topic']
            topics_to_add = []
            if len(row_values) > topic_col and row_values[topic_col]:
                topic_val = row_values[topic_col]
                if 'TOPIC' not in topic_val.upper() or len(topic_val) > 30:
                    if len(topic_val) > 3 and topic_val.upper() != 'NONE':
                        topics_to_add.append(topic_val)
            
            # Initialize unit
            if current_unit not in units:
                unit_name = current_unit_name if current_unit_name else f"Unit {current_unit}"
                units[current_unit] = {
                    "unit_number": current_unit,
                    "unit_name": unit_name,
                    "part1_topics": [],  # List of {"topic": str, "subtopics": [str]}
                    "part2_topics": []   # List of {"topic": str, "subtopics": [str]}
                }
            elif current_unit_name and units[current_unit]["unit_name"] == f"Unit {current_unit}":
                units[current_unit]["unit_name"] = current_unit_name
            
            # Add topics (with subtopics)
            for topic_text in topics_to_add:
                if topic_text:
                    # Get subtopic if available
                    subtopics = []
                    subtopic_col = col_indices.get('subtopic', 5)
                    if len(row_values) > subtopic_col and row_values[subtopic_col]:
                        subtopic_val = row_values[subtopic_col].strip()
                        if subtopic_val and len(subtopic_val) > 3 and 'SUB' not in subtopic_val.upper()[:10]:
                            subtopics.append(subtopic_val)
                    
                    target_part = part_num if part_num else 1
                    target_list = units[current_unit]["part1_topics"] if target_part == 1 else units[current_unit]["part2_topics"]
                    
                    # Check if topic already exists
                    existing_topic = next((t for t in target_list if t["topic"] == topic_text), None)
                    if existing_topic:
                        for st in subtopics:
                            if st not in existing_topic["subtopics"]:
                                existing_topic["subtopics"].append(st)
                    else:
                        target_list.append({"topic": topic_text, "subtopics": subtopics})
        
        return list(units.values())
    
    def _parse_docx(self, file_path: str) -> List[Dict]:
        """Parse DOCX CDAP"""
        try:
            doc = Document(file_path)
            full_text = "\n".join([para.text for para in doc.paragraphs])
            
            # Extract from tables
            for table in doc.tables:
                for row in table.rows:
                    row_text = " | ".join([cell.text.strip() for cell in row.cells if cell.text.strip()])
                    if row_text:
                        full_text += "\n" + row_text
            
            return self._extract_cdap_structure(full_text)
        except Exception as e:
            print(f"DOCX parse error: {e}")
            return []
    
    def _parse_excel(self, file_path: str) -> List[Dict]:
        """Parse Excel CDAP - expects columns: Unit, Part, Topic or similar
        
        Expected structure (e.g., OS CDAP.xlsx):
        - Column 0 (A): UNIT number (1.0, 2.0, etc.)
        - Column 1 (B): SYLLABUS (UNIT NAME)
        - Column 2 (C): OUTCOME (CO)
        - Column 3 (D): PART NO. (1.0 or 2.0)
        - Column 4 (E): TOPICS TO BE COVERED
        - Column 5 (F): SUB TOPICS
        - Column 6 (G): BT LEVEL
        """
        try:
            wb = openpyxl.load_workbook(file_path)
            ws = wb.active
            
            units = {}
            current_unit = None
            current_unit_name = None
            
            # Find the header row (skip rows with "#REF!" or empty)
            header_row_idx = 1
            for row_idx, row in enumerate(ws.iter_rows(min_row=1, max_row=5, values_only=True), start=1):
                row_values = [str(c).strip().upper() if c else '' for c in row]
                # Look for 'UNIT' or 'PART' in header
                if any('UNIT' in v or 'PART' in v or 'TOPIC' in v for v in row_values):
                    header_row_idx = row_idx
                    break
            
            # Detect column indices from header
            col_indices = {'unit': 0, 'unit_name': 1, 'part': 3, 'topic': 4, 'subtopic': 5}
            header = list(ws.iter_rows(min_row=header_row_idx, max_row=header_row_idx, values_only=True))[0]
            for idx, cell in enumerate(header):
                if cell:
                    cell_upper = str(cell).upper()
                    if 'UNIT' in cell_upper and 'NAME' not in cell_upper and idx < 2:
                        col_indices['unit'] = idx
                    elif 'SYLLABUS' in cell_upper or ('UNIT' in cell_upper and 'NAME' in cell_upper):
                        col_indices['unit_name'] = idx
                    elif 'PART' in cell_upper:
                        col_indices['part'] = idx
                    elif 'TOPIC' in cell_upper and 'SUB' not in cell_upper:
                        col_indices['topic'] = idx
                    elif 'SUB' in cell_upper and 'TOPIC' in cell_upper:
                        col_indices['subtopic'] = idx
            
            # Parse data rows (skip header)
            for row in ws.iter_rows(min_row=header_row_idx + 1, values_only=True):
                if not row or all(cell is None for cell in row):
                    continue
                
                row_values = [str(c).strip() if c is not None else '' for c in row]
                
                # Skip rows with #REF! or invalid data
                if '#REF!' in row_values[0] if row_values else False:
                    continue
                
                # Initialize variables for this row
                part_num = None
                topic = None
                
                # Check column for Unit Number (col 0 by default)
                unit_col = col_indices['unit']
                if len(row_values) > unit_col and row_values[unit_col]:
                    unit_val = row_values[unit_col]
                    # Skip header values
                    if unit_val.upper() not in ['UNIT', 'UNIT NO', 'UNIT NO.', 'UNIT NUMBER']:
                        # Match "Unit 1", "1", "1.0", etc.
                        u_match = re.search(r'(?:unit|module)?\s*[:\-]?\s*([\d.]+)', unit_val, re.IGNORECASE)
                        if u_match:
                            unit_num = int(float(u_match.group(1)))
                            current_unit = unit_num
                            # Get unit name from unit_name column
                            unit_name_col = col_indices['unit_name']
                            if len(row_values) > unit_name_col and row_values[unit_name_col]:
                                current_unit_name = row_values[unit_name_col]
                
                if current_unit is None:
                    current_unit = 1  # Default
                    
                # Get Part Number from the specific column (col 3 by default)
                part_col = col_indices['part']
                if len(row_values) > part_col and row_values[part_col]:
                    part_val = row_values[part_col]
                    if part_val.upper() not in ['PART NO', 'PART NO.', 'PART NUMBER', 'PART']:
                        if part_val in ['1', '1.0', 'Part 1', 'Part-1', 'I', 'PART I']:
                            part_num = 1
                        elif part_val in ['2', '2.0', 'Part 2', 'Part-2', 'II', 'PART II']:
                            part_num = 2
                        else:
                            # Try to extract number
                            p_match = re.search(r'[\d.]+', part_val)
                            if p_match:
                                part_num = int(float(p_match.group()))
                
                # Get Topic from specific column (col 4 by default) and optionally subtopic
                topic_col = col_indices['topic']
                subtopic_col = col_indices['subtopic']
                
                topics_to_add = []
                if len(row_values) > topic_col and row_values[topic_col]:
                    topic_val = row_values[topic_col]
                    # Skip header values
                    if 'TOPIC' not in topic_val.upper() or len(topic_val) > 30:
                        if len(topic_val) > 3:
                            topics_to_add.append(topic_val)
                
                # Initialize unit dict if needed
                if current_unit not in units:
                    unit_name = current_unit_name if current_unit_name else f"Unit {current_unit}"
                    units[current_unit] = {
                        "unit_number": current_unit,
                        "unit_name": unit_name,
                        "part1_topics": [],  # List of {"topic": str, "subtopics": [str]}
                        "part2_topics": []   # List of {"topic": str, "subtopics": [str]}
                    }
                elif current_unit_name and units[current_unit]["unit_name"] == f"Unit {current_unit}":
                    # Update unit name if we found a better one
                    units[current_unit]["unit_name"] = current_unit_name
                
                # Add topics to the appropriate part (with subtopics)
                for topic_text in topics_to_add:
                    if topic_text:
                        # Get subtopic if available
                        subtopics = []
                        if len(row_values) > subtopic_col and row_values[subtopic_col]:
                            subtopic_val = row_values[subtopic_col].strip()
                            if subtopic_val and len(subtopic_val) > 3 and 'SUB' not in subtopic_val.upper()[:10]:
                                subtopics.append(subtopic_val)
                        
                        # Default to Part 1 if no part number found
                        target_part = part_num if part_num else 1
                        target_list = units[current_unit]["part1_topics"] if target_part == 1 else units[current_unit]["part2_topics"]
                        
                        # Check if topic already exists, if so, add subtopic to it
                        existing_topic = next((t for t in target_list if t["topic"] == topic_text), None)
                        if existing_topic:
                            for st in subtopics:
                                if st not in existing_topic["subtopics"]:
                                    existing_topic["subtopics"].append(st)
                        else:
                            target_list.append({"topic": topic_text, "subtopics": subtopics})
            
            return list(units.values())
        except Exception as e:
            print(f"Excel parse error: {e}")
            import traceback
            traceback.print_exc()
            return []
    
    def _extract_cdap_structure(self, text: str) -> List[Dict]:
        """Extract CDAP structure from text - identifies units and Part 1/Part 2 sections"""
        units = []
        current_unit = None
        current_part = None
        
        lines = text.split('\n')
        
        for line in lines:
            line = line.strip()
            if not line:
                continue
            
            # Detect unit/module headers
            # Pattern 1: "Unit 1", "Module 1"
            unit_match = re.search(
                r'(?:unit|module)\s*[-:]?\s*([IVX\d]+)(?:\s*[-:]\s*(.+))?',
                line, re.IGNORECASE
            )
            # Pattern 2: Start of line "1 Introduction...", "2 System..." or "1 | Introduction..." (table output)
            if not unit_match:
                # Allow space OR pipe as separator
                simple_match = re.match(r'^(\d+)(?:\s+|(?:\s*[|]\s*))([A-Za-z].+)', line)
                if simple_match:
                    # Check if it looks like a unit header (short number, long title)
                    u_num = int(simple_match.group(1))
                    # Basic sanity: Unit numbers are usually 1-10
                    if 1 <= u_num <= 10: 
                        unit_match = simple_match

            if unit_match:
                unit_num = self._roman_to_int(unit_match.group(1))
                unit_name = unit_match.group(2).strip() if unit_match.group(2) else f"Unit {unit_num}"
                
                current_unit = {
                    "unit_number": unit_num,
                    "unit_name": unit_name,
                    "part1_topics": [],  # List of {"topic": str, "subtopics": [str]}
                    "part2_topics": []   # List of {"topic": str, "subtopics": [str]}
                }
                units.append(current_unit)
                continue
            
            # Detect Part 1/Part 2 headers
            part1_match = re.search(r'part\s*[-:]?\s*1|part\s*i(?:\s|$)', line, re.IGNORECASE)
            part2_match = re.search(r'part\s*[-:]?\s*2|part\s*ii(?:\s|$)', line, re.IGNORECASE)
            
            if part1_match:
                current_part = 1
                continue
            elif part2_match:
                current_part = 2
                continue
            
            # Add topics to current part
            if current_unit and current_part and self._is_valid_topic(line):
                topic = self._clean_topic(line)
                if topic:
                    topic_entry = {"topic": topic, "subtopics": []}
                    if current_part == 1:
                        if not any(t["topic"] == topic for t in current_unit["part1_topics"]):
                            current_unit["part1_topics"].append(topic_entry)
                    else:
                        if not any(t["topic"] == topic for t in current_unit["part2_topics"]):
                            current_unit["part2_topics"].append(topic_entry)
        
        return units
    
    def _is_valid_topic(self, text: str) -> bool:
        """Check if text is a valid topic"""
        text = text.strip()
        if len(text) < 5:
            return False
        if re.match(r'^[\d.]+$', text):
            return False
        if text.lower() in ['part 1', 'part 2', 'part i', 'part ii', 'topics']:
            return False
        return True
    
    def _clean_topic(self, text: str) -> str:
        """Clean topic string"""
        # Remove leading numbers/bullets
        text = re.sub(r'^[\d.•\-–]+\s*', '', text)
        # Remove extra whitespace
        text = ' '.join(text.split())
        return text.strip()
    
    def _roman_to_int(self, s: str) -> int:
        """Convert Roman numeral or digit to integer"""
        s = s.strip().upper()
        if s.isdigit():
            return int(s)
        
        roman = {'I': 1, 'V': 5, 'X': 10}
        result = 0
        prev = 0
        for char in reversed(s):
            val = roman.get(char, 0)
            if val < prev:
                result -= val
            else:
                result += val
            prev = val
        return result if result > 0 else 1


cdap_parser = CDAPParser()
